# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/ipc/PContent.ipdl
# Commit: 6ebc7618a5fa
# Full Hash: 6ebc7618a5fadb59585a4d8b9f17eb65e5654cc6
# Author: Anny Gakhokidze <agakhokidze@mozilla.com>
# Date: 2020-03-10 21:44:54
# Regressor Bug: 1546759
# File Overlap Count: 2
# Description:
#   Bug 1546759 - Change nsSHistory::WalkHistoryEntries to walk browsing context tree instead of doc shell tree, r=peterv,nika
#   
#   WalkHistoryEntries function gets called by nsSHistory::CloneAndReplaceChild
#   and nsSHistory::SetChildHistoryEntry recursively, so those have to be moved
#   into the parent process. This eliminates many sync IPC calls.
# ==============================================================================

diff -r ed75364b23c3 -r 6ebc7618a5fa dom/ipc/PContent.ipdl
--- a/dom/ipc/PContent.ipdl	Tue Mar 10 11:39:06 2020 +0000
+++ b/dom/ipc/PContent.ipdl	Tue Mar 10 14:28:22 2020 +0000
@@ -123,6 +123,7 @@
 using mozilla::dom::MaybeMediaMetadataBase from "mozilla/dom/MediaSessionIPCUtils.h";
 using refcounted class nsDocShellLoadState from "nsDocShellLoadState.h";
 using mozilla::dom::ServiceWorkerShutdownState::Progress from "mozilla/dom/ServiceWorkerShutdownState.h";
+using refcounted class mozilla::dom::CrossProcessSHEntry from "mozilla/dom/MaybeNewPSHEntry.h";
 
 union ChromeRegistryItem
 {
@@ -344,12 +345,6 @@
     uint64_t innerWindowId;
 };
 
-union PSHEntryOrSharedID
-{
-  PSHEntry;
-  uint64_t;
-};
-
 struct KeyValuePair
 {
     nsString key;
@@ -874,6 +869,11 @@
 
     async DisplayLoadError(MaybeDiscardedBrowsingContext aContext, nsString aURI);
 
+    // Tell aContext's docshell to update its mOSHE and mLSHE entries
+    async UpdateSHEntriesInDocShell(CrossProcessSHEntry aOldEntry,
+                                    CrossProcessSHEntry aNewEntry,
+                                    MaybeDiscardedBrowsingContext aContext);
+
 parent:
     async InitBackground(Endpoint<PBackgroundParent> aEndpoint);
 
@@ -949,8 +949,8 @@
 
     async PSHistory(MaybeDiscardedBrowsingContext aContext);
 
-    // Clone from entry or use shared id.
-    sync PSHEntry(nullable PSHistory shistory, PSHEntryOrSharedID entryOrSharedID);
+    // Clone using shared id.
+    sync PSHEntry(nullable PSHistory shistory, uint64_t sharedID);
 
     async PBenchmarkStorage();
 
@@ -1538,6 +1538,13 @@
     async ReportServiceWorkerShutdownProgress(uint32_t aShutdownStateId,
                                               Progress aProgress);
 
+    /**
+     * Whenever docshell updates its SH entries, we need to update them in
+     * the corresponding BrowsingContext
+     */
+    async UpdateSHEntriesInBC(nullable PSHEntry aNewLSHE, nullable
+                              PSHEntry aNewOSHE, MaybeDiscardedBrowsingContext aContext);
+
 both:
     async ScriptError(nsString message, nsString sourceName, nsString sourceLine,
                       uint32_t lineNumber, uint32_t colNumber, uint32_t flags,